metals/src/main/scala/scala/meta/internal/metals/WorkspaceLspService.scala

```diff
@@ -0,0 +1,1152 @@
+      treeView.parent(params)
+    }.asJava
+  }
+
+  override def treeViewVisibilityDidChange(
+      params: TreeViewVisibilityDidChangeParams
+  ): CompletableFuture[Unit] =
+    Future {
+      treeView.onVisibilityDidChange(params)
+    }.asJava
+
+  override def treeViewNodeCollapseDidChange(
+      params: TreeViewNodeCollapseDidChangeParams
+  ): CompletableFuture[Unit] =
+    Future {
+      treeView.onCollapseDidChange(params)
+    }.asJava
+
+  override def treeViewReveal(
+      params: TextDocumentPositionParams
+  ): CompletableFuture[TreeViewNodeRevealResult] =
+    Future {
+      treeView
+        .reveal(
+          params.getTextDocument().getUri().toAbsolutePath,
+          params.getPosition(),
+        )
+        .orNull
+    }.asJava
+
+  override def findTextInDependencyJars(
+      params: FindTextInDependencyJarsRequest
+  ): CompletableFuture[ju.List[Location]] =
+    collectSeq(_.findTextInDependencyJars(params))(_.flatten.asJava).asJava
+
+  override def doctorVisibilityDidChange(
+      params: DoctorVisibilityDidChangeParams
+  ): CompletableFuture[Unit] =
+    onCurrentFolder(
+      _.doctorVisibilityDidChange(params),
+      "change doctor visibility",
+    ).asJava
+
+  override def didFocus(
+      params: AnyRef
+  ): CompletableFuture[DidFocusResult.Value] = {
+    val uriOpt: Option[String] = params match {
+      case string: String =>
+        Option(string)
+      case (h: String) :: Nil =>
+        Option(h)
+      case _ =>
+        scribe.warn(
+          s"Unexpected notification params received for didFocusTextDocument: $params"
+        )
+        None
+    }
+    uriOpt match {
+      case Some(uri) =>
+        focusedDocument = Some(uri.toAbsolutePath)
+        getServiceFor(uri).didFocus(uri)
+      case None =>
+        CompletableFuture.completedFuture(DidFocusResult.NoBuildTarget)
+    }
+  }
+
+  override def windowStateDidChange(params: WindowStateDidChangeParams): Unit =
+    if (params.focused) {
+      folderServices.foreach(_.unpause())
+    } else {
+      folderServices.foreach(_.pause())
+    }
+
+  private def notFound(objectName: String, id: String) =
+    languageClient.showMessage(
+      Messages.errorMessageParams(s"$objectName not found: $id")
+    )
+
+  private def definedInMultipleWorkspaceFolders(
+      objectName: String,
+      id: String,
+  ) =
+    languageClient.showMessage(
+      Messages.errorMessageParams(
+        s"$objectName defined in multiple workspace folders: $id"
+      )
+    )
+
+  override def executeCommand(
+      params: ExecuteCommandParams
+  ): CompletableFuture[Object] =
+    params match {
+      case ServerCommands.ScanWorkspaceSources() =>
+        foreachSeq(_.indexSources(), ignoreValue = true)
+      case ServerCommands.RestartBuildServer() =>
+        folderServices.find(_.isBloop()).foreach(_.shutDownBloop())
+        foreachSeq(_.autoConnectToBuildServer())
+      case ServerCommands.GenerateBspConfig() =>
+        foreachSeq(
```

Comment:
This could start multiple sbt processes, we should only do it for the focused folder, same with the below commands related to build server.